Declarative mission specification

zaq32 opened the pull request:
This change introduces mission loop coordinator with list of update/verify/action tasks specified in a completely declarative way. All mission tasks are specified as part of the mission loop variadic template parameter pack. Mission loop coordinator in turn recognizes task types, classifies and extracts the descriptors from them, and then stores them in type specific lists whose sizes are calculated to match the necessary requirements.  Majority of the work is done at compile time while the remaining (much smaller part) is done during initialization. 

Additional changes: 
* Mission state is not longer fixed. and can be changed/replaced as long as all of the actions agree on it. 
* Extract all of the implicit mission tasks into standalone libraries (sail, time, adcs).
* Minor doxygen configuration update.
* Add possibility to suspend/resume mission processing from the terminal and run mission in manual mode.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pw-sat2/pwsat2obc/58)
<!-- Reviewable:end -->

Novakov commented:
Reviewed 53 of 53 files at r1, 4 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[integration_tests/tests/test_antenna_deployment.py, line 56 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYz1I0eFMr0nwdoBB0G:-KYz1I0eFMr0nwdoBB0H:b-o11eq0) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/integration_tests/tests/test_antenna_deployment.py#L56)):*
> ```Python
>         self.system.backup_antenna.on_deployment_cancel = event.set
>         self.begin()
>         self.next_step()
> ```

Add method `next_steps(steps_count)` to it will be easier to invoke N steps at once

---

*[integration_tests/tests/test_antenna_deployment.py, line 58 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYz1D7gsmpxKs1MWoTV:-KYz1D7gsmpxKs1MWoTW:bnds8pp) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/integration_tests/tests/test_antenna_deployment.py#L58)):*
> ```Python
>         self.next_step()
>         self.next_step()
>         self.next_step()
> ```

I can't think of better solution right now but I don't like the idea of invoking arbitrary number of steps to check for some action - it makes tests harder to mantain in case of changes

---

*[libs/mission/Include/mission/base.hpp, line 86 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYzCDvAdps7Ub3AussM:-KYzCDvAdps7Ub3AussN:br00xve) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/libs/mission/Include/mission/base.hpp#L86)):*
> ```cpp
>         UpdateWarning,
>         /** @brief Fatal error during update */
>         UpdateFailure
> ```

Just `OK`. `Warning` and `Failure` would be ok

---

*[libs/mission/Include/mission/base.hpp, line 131 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYzCWrzaI043Ev8trSI:-KYzCWrzaI043Ev8trSJ:b56wl4w) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/libs/mission/Include/mission/base.hpp#L131)):*
> ```cpp
>         VerifyOK,
>         /** @brief State is invalid. */
>         VerifyFailure
> ```

Just `OK` and `Failure`

---

*[libs/mission/sail/include/mission/sail.hpp, line 38 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYz3SbQpJo0JqoSRaOg:-KYz3SbQpJo0JqoSRaOh:b-ju7fze) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/libs/mission/sail/include/mission/sail.hpp#L38)):*
> ```cpp
>          * @param[in] initial Initial sail deployment state.
>          */
>         SailTask(bool initial);
> ```

I'm naming-nazi. Name `initial` says more or less nothing. It's better to name it like `isSailOpened`.

Also the 'sail deployment state' is mentioned in few places here, maybe it will be wise to define in as enum with two members?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[integration_tests/tests/test_antenna_deployment.py, line 56 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYz1I0eFMr0nwdoBB0G:-KZ2wTUg5JkEnY1x6_WS:b-886gsx) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/integration_tests/tests/test_antenna_deployment.py#L56)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

Add method `next_steps(steps_count)` to it will be easier to invoke N steps at once
</blockquote></details>

Fixed

---

*[integration_tests/tests/test_antenna_deployment.py, line 58 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYz1D7gsmpxKs1MWoTV:-KZ2wUCoymuTOEnlPbTD:b-gir8r7) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/integration_tests/tests/test_antenna_deployment.py#L58)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

I can't think of better solution right now but I don't like the idea of invoking arbitrary number of steps to check for some action - it makes tests harder to mantain in case of changes
</blockquote></details>

I also don't like it but for the moment I don't have any better solutions

---

*[libs/mission/Include/mission/base.hpp, line 86 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYzCDvAdps7Ub3AussM:-KZ2qJWmrcgLc8T2FVyX:b-886gsx) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/libs/mission/Include/mission/base.hpp#L86)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

Just `OK`. `Warning` and `Failure` would be ok
</blockquote></details>

Fixed

---

*[libs/mission/Include/mission/base.hpp, line 131 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYzCWrzaI043Ev8trSI:-KZ2qIugeAay099QSF6x:b-886gsx) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/libs/mission/Include/mission/base.hpp#L131)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

Just `OK` and `Failure`
</blockquote></details>

Fixed

---

*[libs/mission/sail/include/mission/sail.hpp, line 38 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58#-KYz3SbQpJo0JqoSRaOg:-KZ2qGrRRgDd28ix3A-B:b-886gsx) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/d16f9460a51a0a356f49f4fabc440425f87e080c/libs/mission/sail/include/mission/sail.hpp#L38)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

I'm naming-nazi. Name `initial` says more or less nothing. It's better to name it like `isSailOpened`.

Also the 'sail deployment state' is mentioned in few places here, maybe it will be wise to define in as enum with two members?
</blockquote></details>

Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58)*
<!-- Sent from Reviewable.io -->

Novakov commented:
Reviewed 12 of 12 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/58)*
<!-- Sent from Reviewable.io -->
